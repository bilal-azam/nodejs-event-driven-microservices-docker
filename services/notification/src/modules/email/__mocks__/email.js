/* eslint-disable no-unused-vars */
const email = {};

email.sendArticleAddedEmail = jest.fn((message) => {
});

module.exports = email;
