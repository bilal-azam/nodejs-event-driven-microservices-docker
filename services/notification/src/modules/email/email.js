const nodemailer = require('nodemailer');
const logger = require('winston');
const config = require('../../environment/config');
const template = require('./email.templates');

const transporter = nodemailer.createTransport({
  service: config.email.service,
  auth: {
    user: config.email.username,
    pass: config.email.password,
  },
});

const email = {};

email.sendArticleAddedEmail = async (message) => {
  try {
    const emailHtml = template.GetRenderedArticleAddedEmailHtml(message.title, message.description);

    const mailOptions = {
      from: config.email.username,
      to: config.email.adminEmailID,
      subject: 'LocalNewsApplication - New Article Added!',
      html: emailHtml,
    };
    await transporter.sendMail(mailOptions);
  } catch (err) {
    logger.error(`Error sending email ${err}`);
  }
};

module.exports = email;
