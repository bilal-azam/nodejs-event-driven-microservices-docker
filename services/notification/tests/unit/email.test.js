/* eslint-disable no-unused-vars */
/* eslint-disable global-require */
jest.mock('nodemailer');
jest.mock('winston');
jest.mock('../../src/modules/email/email.templates');

const nodemailer = require('nodemailer');
const winston = require('winston');
const templates = require('../../src/modules/email/email.templates');

const message = {
  title: 'Test Article',
  description: 'Test Text',
};

describe('Email Options', () => {
  test('Shoud set the correct email credentials', () => {
    require('../../src/modules/email/email');
    expect(nodemailer.createTransport).toBeCalledTimes(1);
    expect(nodemailer.createTransport.mock.calls[0][0]).toMatchSnapshot();
    expect(winston.error).not.toBeCalled();
  });
});

describe('Send Article Added Email', () => {
  afterEach(() => {
    templates.GetRenderedArticleAddedEmailHtml.mockClear();
  });

  test('Shoud send email when correct details are provided', () => {
    const email = require('../../src/modules/email/email');

    email.sendArticleAddedEmail(message);

    expect(templates.GetRenderedArticleAddedEmailHtml).toBeCalledTimes(1);
    expect(templates.GetRenderedArticleAddedEmailHtml)
      .toBeCalledWith(message.title, message.description);

    expect(nodemailer.sendMail.mock.calls.length).toBe(1);
    expect(nodemailer.sendMail.mock.calls[0][0]).toMatchSnapshot();
  });

  test('Shoud throw an error when an exception occurs', () => {
    templates.GetRenderedArticleAddedEmailHtml = jest.fn((title, description) => {
      throw new Error();
    });

    const email = require('../../src/modules/email/email');

    email.sendArticleAddedEmail(message);

    expect(templates.GetRenderedArticleAddedEmailHtml).toBeCalledTimes(1);
    expect(templates.GetRenderedArticleAddedEmailHtml)
      .toBeCalledWith(message.title, message.description);

    expect(nodemailer.sendMail.mock.calls.length).toBe(1);
    expect(nodemailer.sendMail.mock.calls[0][0]).toMatchSnapshot();

    expect(winston.error).toBeCalledTimes(1);
  });
});
