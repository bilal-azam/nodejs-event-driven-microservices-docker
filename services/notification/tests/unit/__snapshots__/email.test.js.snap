// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email Options Shoud set the correct email credentials 1`] = `
Object {
  "auth": Object {
    "pass": "",
    "user": "",
  },
  "service": "",
}
`;

exports[`Send Article Added Email Shoud send email when correct details are provided 1`] = `
Object {
  "from": "",
  "html": "Test Article, Test Text",
  "subject": "LocalNewsApplication - New Article Added!",
  "to": "",
}
`;

exports[`Send Article Added Email Shoud throw an error when an exception occurs 1`] = `
Object {
  "from": "",
  "html": "Test Article, Test Text",
  "subject": "LocalNewsApplication - New Article Added!",
  "to": "",
}
`;
